### ISSUE TITLE
different token length of input and output image embeddings

### ISSUE THREAD

**Issue by jiangzhengguo** [modelscope/Nexus-Gen#16]
I want to express my appreciation for the excellent work! The results in multimodal understanding and generation are truly impressive, especially achieving such remarkable performance on such a complex task. It's really inspiring!

I have a small question regarding the dataset construction. While reviewing the project, I noticed that the input image token count is 324, whereas the output token count is 81. However, the paper mentions that the input and output share the same embedding space, which has raised a bit of confusion for me.
My guess is that a mean pooling operation might have been applied to the input images. Could you confirm if this is the case, or if another method was used?

**Comment by mi804** [comment]
Thank you for your recognition of Nexus-Gen. Nexus-Gen adopts a unified embedding space based on the ViT embedding space of Qwen2.5-VL. Due to Qwen2.5-V's dynamic resolution capability, images of different sizes will yield varying numbers of tokens after encoding. Through experimentation, we found that the llm struggles to accurately predict as many as 324 image tokens, so we have limited the output token count to 81, corresponding to an image resolution of 252x252. In reality, both the input 324 tokens and the output 81 image tokens still reside within the same feature space.

We will also be rolling out a general update for Nexus-Gen in the near future, significantly improving its scores in image generation and comprehension capabilities.

**Comment by fourierer** [comment]
> 以下是用于微调 llm 的一个小示例 jsonl 数据集：
> 
> ```
> {"id": "EliGen_920035", "source": "EliGen", "messages": [{"role": "user", "content": "Generate an image according to the following description: A breathtaking sci-fi scene features a fleet of futuristic flying saucers navigating through a rugged, mountainous landscape."}, {"role": "assistant", "content": "Here is an image based on the description: <image>"}], "images": ["data/images/920035.png"]}
> {"id": "EliGen_920041", "source": "EliGen", "messages": [{"role": "user", "content": "Generate an image according to the following description: A young woman with dark, intricately styled hair adorned with pink roses and white flowers is the main subject of this image."}, {"role": "assistant", "content": "Here is an image based on the description: <image>"}], "images": ["data/images/920041.png"]}
> ```
> 
> 下面是将其转换为嵌入数据集以训练解码器的脚本
> 
> ```
> import pandas as pd
> import torch
> import torch.multiprocessing as mp
> import random
> import os
> from tqdm import tqdm
> import logging
> import json
> from modelscope import snapshot_download
> from transformers import Qwen2_5_VLForConditionalGeneration, AutoProcessor
> from torchvision import transforms
> from PIL import Image
> 
> 
> # Configure logging
> logging.basicConfig(level=logging.INFO, format='%(asctime)s [Process %(processName)s] %(levelname)s: %(message)s')
> 
> 
> def get_image_embeds(rank, samples, out_embed_dir, output_jsonl, lock, height=252):
>     torch.cuda.set_device(rank)
>     logging.info(f"Process {os.getpid()} using device {torch.cuda.current_device()}: {torch.cuda.get_device_name()}")
>     path = snapshot_download("Qwen/Qwen2.5-VL-7B-Instruct")
>     model = Qwen2_5_VLForConditionalGeneration.from_pretrained(
>         path, torch_dtype="auto", device_map=f"cuda:{rank}"
>     )
>     model.eval()
>     visual_model = model.visual
>     processor = AutoProcessor.from_pretrained("Qwen/Qwen2.5-VL-7B-Instruct")
>     for sample in tqdm(samples, desc=f"Process {rank}", position=rank):
>         try:
>             image_path = sample['images'][0]
>             image_id = sample['id']
>             embed_path = os.path.join(out_embed_dir, f"{image_id}.pt")
>             sample['embed_path'] = embed_path
>             sample['image'] = [image_path]
> 
>             # transform image
>             image = Image.open(image_path).convert('RGB')
>             img_width, img_height = image.size
>             scale = max(height / img_width, height / img_height)
>             shape = [round(img_height * scale), round(img_width * scale)]
>             image_scaled = transforms.functional.resize(image, shape, interpolation=transforms.InterpolationMode.BILINEAR)
>             img_transform = transforms.CenterCrop((height, height))
>             image = img_transform(image_scaled)
> 
>             # Process image
>             media_inputs = processor.image_processor(images=[image], videos=None, return_tensors='pt', do_resize=False)
>             pixel_values = media_inputs["pixel_values"].to(visual_model.device)
>             image_grid_thw = media_inputs["image_grid_thw"].to(visual_model.device)
>             pixel_values = pixel_values.type(visual_model.dtype)
>             with torch.no_grad():
>                 image_embeds = visual_model(pixel_values, grid_thw=image_grid_thw).cpu()
>                 # (1, 81, 3584)
>             torch.save(image_embeds, embed_path)
> 
>             # Lock the writing to output_jsonl
>             with lock:
>                 with open(output_jsonl, 'a') as out_file:
>                     out_file.write(json.dumps(sample) + '\n')
>         except Exception as e:
>             logging.error(f"Error processing image {image_path}: {e}")
> 
> 
> if __name__ == "__main__":
>     height = 252 # with respect to 81 tokens
>     output_folder = f"data/embed_dataset"
>     input_jsonl = "data/test.jsonl"
>     output_jsonl = os.path.join(output_folder, f"image_embeds_{height}.jsonl")
>     out_embed_dir = os.path.join(output_folder, f"embeds_{height}")
> 
>     os.makedirs(output_folder, exist_ok=True)
>     os.makedirs(out_embed_dir, exist_ok=True)
>     image_field = 'id'
> 
>     processed_images = set()
>     print('reading processed images...')
>     if os.path.exists(output_jsonl):
>         with open(output_jsonl, 'r') as f:
>             for line in f:
>                 processed_images.add(json.loads(line)[image_field])
>     print(f'finished reading {len(processed_images)} processed images')
> 
>     samples = []
>     reading_limit = 100000000
>     with open(input_jsonl, 'r') as f:
>         for line in tqdm(f):
>             sample = json.loads(line)
>             image_id = sample[image_field]
>             if image_id in processed_images:
>                 continue
>             samples.append(sample)
>             if len(samples) % reading_limit == 0:
>                 break
>     print(f"reading {len(samples)} new images")
> 
>     available_gpus = torch.cuda.device_count()
>     num_processes = min(8, available_gpus)
>     mp.set_start_method('spawn', force=True)
> 
>     # Create a lock for file writing synchronization
>     manager = mp.Manager()
>     lock = manager.Lock()
> 
>     processes = []
>     total_missing = len(samples)
>     chunk_size = total_missing // num_processes
>     for rank in range(num_processes):
>         process_start = rank * chunk_size
>         process_end = (rank + 1) * chunk_size if rank < num_processes - 1 else total_missing
>         p = mp.Process(target=get_image_embeds, args=(rank, samples[process_start:process_end], out_embed_dir, output_jsonl, lock, height))
>         p.start()
>         processes.append(p)
> 
>     for p in processes:
>         p.join()
> ```

hi, could you an example jsonl for image editing? Are these keys ('id', 'source') necessary in jsonl?

**Comment by mi804** [comment]
Here is an example:
```
{"id": "ImagePulseFaceID_1744652050482332924_1744675068430312018_1", "source": "ImagePulseFaceID", "images": ["data/edit_dataset/imagepulse/FaceID/images/1744652050482332924/1744675068041601686.png", "data/edit_dataset/imagepulse/FaceID/images/1744652050482332924/1744675067673353994.png"], "messages": [{"role": "user", "content": "<image> Change the woman's outfit to a dark, ornate dress with a long black cape and change the background to a mystical, glowing scene."}, {"role": "assistant", "content": "Here is the image: <image>"}]}
```
'id', 'source' are just identification for the sample, and are not necessary for training.

**Comment by fourierer** [comment]
> Here is an example:
> 
> ```
> {"id": "ImagePulseFaceID_1744652050482332924_1744675068430312018_1", "source": "ImagePulseFaceID", "images": ["data/edit_dataset/imagepulse/FaceID/images/1744652050482332924/1744675068041601686.png", "data/edit_dataset/imagepulse/FaceID/images/1744652050482332924/1744675067673353994.png"], "messages": [{"role": "user", "content": "<image> Change the woman's outfit to a dark, ornate dress with a long black cape and change the background to a mystical, glowing scene."}, {"role": "assistant", "content": "Here is the image: <image>"}]}
> ```
> 
> 'id', 'source' are just identification for the sample, and are not necessary for training.

Thanks for your reply! Here are the two additional questions i'd like to realize:
(1) In the method you provided for generating embeddings to train the decoder, you used the vision encoder from Qwen2.5-VL-7B-Instruct to generate embeddings, and then reconstructed the original image using these embeddings. However, in the image_editing.py file you provided, it seems that the output of an autoregressive transformer combined with a vision projector is used to reconstruct the image. Is my understanding correct?
(2) The image embedding output by the autoregressive transformer+vision projector is used to compute the MSE and cosine similarity loss with which other embedding? If it is compared with the image embedding output by the vision encoder, then I would understand question (1).

**Comment by mi804** [comment]
Yes, autoregressive transformer+vision projector (The VL model) is train to predict to image embedding output by the vision encoder. The core idea of Nexus-Gen is to connect the VL model and the diffusion decoder in the image embedding space of the vision projector.

**Comment by mi804** [comment]
Simply speaking, VL model's goal is to predict the target image's embedding, which is the last image in image generation and editing tasks. Besides, loss calculation is only performed on the image embeddings in the assistant's response.

So in editing task, image embedding1 is only used to generate response, not used to to calculate loss. The loss is calculated between image embedding2 outputed by the autoregressive transformer (predicted) and the outputed by the vision encoder (groundtruth). Thus, VL model is trained to predict the embeddings of image 2. And decoder will use the predicted embeddings to reconstruct image 2.

**Comment by ODRLver** [comment]
> Yes, autoregressive transformer+vision projector (The VL model) is trained to predict the image embedding output by the vision encoder. The core idea of Nexus-Gen is to connect the VL model and the diffusion decoder in the image embedding space of the vision projector.

Hi, I used the scripts you provided above to generate the embeddings to train the decoder. I encountered some issue when trying to train the decoder, I noticed that you import lighting as pl in flux_trainer.py, is that an old version of pytorch_lightning? Could you provide the version of the pytorch lightning you use? Thanks in advance.
